File: example/java/spring-boot/one_consumer_one_provider/client-api/src/test/java/br/com/zup/pact/client/resource/ClientResourceEndpointTest.java
package br.com.zup.pact.client.resource;

import static org.hamcrest.Matchers.hasSize;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import br.com.zup.pact.client.dto.BalanceDTO;
import br.com.zup.pact.client.dto.ClientDetailsDTO;
import br.com.zup.pact.client.entity.Client;
import br.com.zup.pact.client.service.ClientService;
import br.com.zup.pact.client.stub.ClientStub;
import java.math.BigDecimal;
import java.util.Optional;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.web.servlet.MockMvc;

@SpringBootTest
@AutoConfigureMockMvc
class ClientResourceEndpointTest {

    @Autowired
    private MockMvc mockMvc;

    @MockBean
    private ClientService clientService;

    @Autowired
    private ClientStub clientStub;

    @Test
    void getClientDetails() throws Exception {
        final ClientDetailsDTO clientDetailsDTO = Client.fromEntityToDto(clientStub.getClients().get(1));
        when(clientService.getClientDetails(anyInt())).thenReturn(Optional.of(clientDetailsDTO));
        mockMvc.perform(get("/v1/clients/1"))
                .andDo(print())
                .andExpect(jsonPath("$.id").exists())
                .andExpect(jsonPath("$.name").exists())
                .andExpect(jsonPath("$.finalName").exists())
                .andExpect(jsonPath("$.age").exists())
                .andExpect(jsonPath("$.id").value(clientDetailsDTO.getId()))
                .andExpect(jsonPath("$.name").value(clientDetailsDTO.getName()))
                .andExpect(jsonPath("$.finalName").value(clientDetailsDTO.getFinalName()))
                .andExpect(jsonPath("$.age").value(clientDetailsDTO.getAge()))
                .andExpect(status().isOk());
    }

    @Test
    void getAccountDetailsByNonExistentClientId() throws Exception {
        mockMvc.perform(get("/v1/clients/1000"))
                .andDo(print())
                .andExpect(status().isNotFound());
    }

    @Test
    void getAll() throws Exception {
        when(clientService.getAll()).thenReturn(Optional.of(clientStub.getAllStubsDTOFormat()));
        final ClientDetailsDTO firstClientDetailsDTO = clientStub.getAllStubsDTOFormat().get(0);
        mockMvc.perform(get("/v1/clients"))
                .andDo(print())
                .andExpect(jsonPath("$", hasSize(clientStub.getAllStubsDTOFormat().size())))
                .andExpect(jsonPath("$.[0].id").exists())
                .andExpect(jsonPath("$.[0].name").exists())
                .andExpect(jsonPath("$.[0].finalName").exists())
                .andExpect(jsonPath("$.[0].age").exists())
                .andExpect(jsonPath("$.[0].id").value(firstClientDetailsDTO.getId()))
                .andExpect(jsonPath("$.[0].name").value(firstClientDetailsDTO.getName()))
                .andExpect(jsonPath("$.[0].finalName").value(firstClientDetailsDTO.getFinalName()))
                .andExpect(jsonPath("$.[0].age").value(firstClientDetailsDTO.getAge()))
                .andExpect(status().isOk());
    }

    @Test
    void getAllNoContent() throws Exception {
        when(clientService.getAll()).thenReturn(Optional.empty());
        mockMvc.perform(get("/v1/clients"))
                .andDo(print())
                .andExpect(status().isNotFound());
    }

    @Test
    void getBalance() throws Exception {
        BalanceDTO balanceDTO = buildBalanceDTO();
        when(clientService.getBalance(anyInt())).thenReturn(Optional.of(balanceDTO));
        mockMvc.perform(get("/v1/clients/balance/1"))
                .andDo(print())
                .andExpect(jsonPath("$.clientId").exists())
                .andExpect(jsonPath("$.accountId").exists())
                .andExpect(jsonPath("$.balance").exists())
                .andExpect(jsonPath("$.clientId").value(balanceDTO.getClientId()))
                .andExpect(jsonPath("$.accountId").value(balanceDTO.getAccountId()))
                .andExpect(jsonPath("$.balance").value(balanceDTO.getBalance()))
                .andExpect(status().isOk());
    }

    private BalanceDTO buildBalanceDTO() {
        return BalanceDTO.builder()
                .accountId(1)
                .balance("mil reais")
                .clientId(10)
                .build();
    }
}

File: example/kotlin/one-consumer-one-provider/consumer/src/test/kotlin/br/com/zup/pact/consumer/resource/ClientResourceEndpointTest.kt
package br.com.zup.pact.consumer.resource

import br.com.zup.pact.consumer.dto.BalanceDTO
import br.com.zup.pact.consumer.dto.ClientDetailsDTO
import br.com.zup.pact.consumer.service.ClientService
import io.mockk.every
import io.mockk.mockk
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.TestConfiguration
import org.springframework.context.annotation.Bean
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultHandlers
import org.springframework.test.web.servlet.result.MockMvcResultMatchers

@SpringBootTest
@AutoConfigureMockMvc
class ClientResourceEndpointTest (
        @Autowired val mockMvc: MockMvc,
        @Autowired val clientService: ClientService
) {

    @TestConfiguration
    class ClientResourceEndpointTestConfig {
        @Bean
        fun clientService() = mockk<ClientService>()
    }

    @Test
    fun `Method getAll returns a ClientDetails list`() {
        val clientDetailsDTOList: List<ClientDetailsDTO> = arrayListOf(
                ClientDetailsDTO(1, 1, "any", "any", 40),
                ClientDetailsDTO(2, 2, "any", "any", 25),
                ClientDetailsDTO(3, 3, "any", "any", 20)
        )

        every { clientService.getAll() }
                .returns(clientDetailsDTOList)

        mockMvc.perform(MockMvcRequestBuilders.get("/v1/clients"))
                .andDo(MockMvcResultHandlers.print())
                .andExpect(MockMvcResultMatchers.jsonPath("$.length()").value(clientDetailsDTOList.size))
                .andExpect(MockMvcResultMatchers.jsonPath("$.[0].clientId").isNumber)
                .andExpect(MockMvcResultMatchers.jsonPath("$.[0].accountId").isNumber)
                .andExpect(MockMvcResultMatchers.jsonPath("$.[0].firstName").isString)
                .andExpect(MockMvcResultMatchers.jsonPath("$.[0].lastName").isString)
                .andExpect(MockMvcResultMatchers.jsonPath("$.[0].age").isNumber)
                .andExpect(MockMvcResultMatchers.status().isOk)
    }

    @Test
    fun `Method getAll returns a ClientDetails empty list`() {

        every { clientService.getAll() }
                .returns(emptyList())

        mockMvc.perform(MockMvcRequestBuilders.get("/v1/clients"))
                .andDo(MockMvcResultHandlers.print())
                .andExpect(MockMvcResultMatchers.jsonPath("$.clientId").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.accountId").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.firstName").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.lastName").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.age").doesNotExist())
                .andExpect(MockMvcResultMatchers.status().isNotFound)
    }

    @Test
    fun `Method getClientDetails returns a ClientDetails`() {

        val clientDetailsDTO = ClientDetailsDTO(1,
                1, "any_first_name", "any_last_name", 40)

        every { clientService.getClientDetails(any()) }
                .returns(clientDetailsDTO)

        mockMvc.perform(MockMvcRequestBuilders.get("/v1/clients/1"))
                .andDo(MockMvcResultHandlers.print())
                .andExpect(MockMvcResultMatchers.jsonPath("$.clientId").value(1))
                .andExpect(MockMvcResultMatchers.jsonPath("$.accountId").value(1))
                .andExpect(MockMvcResultMatchers.jsonPath("$.firstName").value("any_first_name"))
                .andExpect(MockMvcResultMatchers.jsonPath("$.lastName").value("any_last_name"))
                .andExpect(MockMvcResultMatchers.jsonPath("$.age").value(40))
                .andExpect(MockMvcResultMatchers.status().isOk)
    }

    @Test
    fun `Method getClientDetails returns not found if client does not exists`() {

        every { clientService.getClientDetails(any()) }
                .returns(null)

        mockMvc.perform(MockMvcRequestBuilders.get("/v1/clients/1"))
                .andDo(MockMvcResultHandlers.print())
                .andExpect(MockMvcResultMatchers.jsonPath("$.clientId").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.accountId").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.firstName").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.lastName").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.age").doesNotExist())
                .andExpect(MockMvcResultMatchers.status().isNotFound)

    }

    @Test
    fun `Method getBalance returns a BalanceDTO`() {

        val balanceDTO = BalanceDTO(1, 1, 250.0)

        every { clientService.getBalance(any()) }
                .returns(balanceDTO)

        mockMvc.perform(MockMvcRequestBuilders.get("/v1/clients/balance/1"))
                .andDo(MockMvcResultHandlers.print())
                .andExpect(MockMvcResultMatchers.jsonPath("$.accountId").exists())
                .andExpect(MockMvcResultMatchers.jsonPath("$.clientId").exists())
                .andExpect(MockMvcResultMatchers.jsonPath("$.balance").exists())
                .andExpect(MockMvcResultMatchers.jsonPath("$.accountId").value(balanceDTO.accountId))
                .andExpect(MockMvcResultMatchers.jsonPath("$.clientId").value(balanceDTO.clientId))
                .andExpect(MockMvcResultMatchers.jsonPath("$.balance").value(balanceDTO.balance))
                .andExpect(MockMvcResultMatchers.status().isOk)

    }

    @Test
    fun `Method getBalance returns not found if balance does not exists`() {

        every { clientService.getBalance(any()) }
                .returns(null)

        mockMvc.perform(MockMvcRequestBuilders.get("/v1/clients/balance/1"))
                .andDo(MockMvcResultHandlers.print())
                .andExpect(MockMvcResultMatchers.jsonPath("$.clientId").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.accountId").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.firstName").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.lastName").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.age").doesNotExist())
                .andExpect(MockMvcResultMatchers.status().isNotFound)

    }
}
